### ISSUE TITLE
Question: zeppelinOS add parameter to initialize()

### ISSUE THREAD

**Issue by JohannesMayerConda** [zeppelinos/zos#417]
Sorry for crossposting but  I've posted this question on stackexchange:
https://ethereum.stackexchange.com/questions/61959/zeppelinos-add-parameter-to-initialize

But it doesn't gather much attention.

When I add a parameter to initialize() I don't know how to update with calling the new parameter...

**Comment by ajsantander** [comment]
Hey @JohannesMayerConda . Sorry for the delayed response and thanks for bringing this up.

The upgraded contract, if it inherits from the old one, will already have an initialize method with two parameters. What you are doing is overloading the new contract with a new initialize method (with the same name as the old one) with three parameters instead of two.

I suspect that the truffle dependency is having trouble identifying the correct method to call, and tries to call the old one. I will confirm if there is an open issue for that and post it here.

On the meantime, as possible workarounds you could:
1) Manually alter the build artifacts of your contract and put the initialize method with three parameters before the other/s.
2) Call your new initialize method something else, like "updatedInitialize(...) initializable".

**Comment by ajsantander** [comment]
@JohannesMayerConda the issue I was referring to was https://github.com/zeppelinos/zos/issues/194

It seems that this has been fixed, so your problem must be something else. Sorry for making you think about something that is NOT your problem!

With this in mind, I suspect that since your proxy has already been initialized via [Initializable.sol](https://github.com/zeppelinos/zos/blob/master/packages/lib/contracts/Initializable.sol), the private state variable "initialized" is set to true, hence you cannot call initialize again in this particular proxy instance.

If you create a new proxy, linked to a logic contract that has the 3 parameters it should work, but not for a proxy that has already been initialized. So, I would (1) add the new initializer with 3 params to your logic contract in case you want to create new proxies and (2) add a separate function to upgrade a proxy with the new foo variable, making sure that this too can only be called once, something like:

```
upgrade(uint _foo) {
  require(foo === 0);
  foo = _foo;
}
``` 

hth!

**Comment by JohannesMayerConda** [comment]
@ajsantander you are right that my proxy has been initialized and cannot be initialized again. You are also right that creating a NEW proxy works.

I think I understand what you mean. I could set new variables in constructor. I find this kind of confusing because I thought I should now avoid using constructors. I think this is at least something that should be documented if this really has to be like this (sorry in case I've overlooked it)...

**Comment by ajsantander** [comment]
No, you're absolutely right. I'll make sure we add it to our docs, thanks!
